cloudinit/net/dhcp.py

```diff
@@ -248,6 +248,8 @@ def dhcp_discovery(dhclient_cmd_path, interface, cleandir):
         else:
             ppid = util.get_proc_ppid(pid)
             if ppid == 1:
+                if len(pout) == 2:
```

Comment:
Why is it here that you want to emit the debugging information, rather than elsewhere?

Reply:
To make sure we capture as much as we can before we kill the dhclient and this is the last point before we return in the success path.

Reply:
I have some reservations about this being in the successful path: this means that we'll be logging a lot more information in this codepath, which will almost always be irrelevant (because everything worked fine).

Can you detail why you want this in the successful path, not in the failed path? (Perhaps describe at a higher-level the problem you're seeing, and why this is a solution to it.)

Reply:
@OddBloke What we have been seeing from some of our investigation into some of the latency issue during boot is slow DHCP response from DHCP server. The output of dhclient's stderr was the one that helped nailed it. In particular, we could see multiple DHCPREQUEST from dhclient (due to DHCP server being overloaded and not responded timely to the first DHCPREQUEST). In addition, we also saw cases where DHCPOFFER was posted, but then DHCPACK arrived too late, which triggered dhclient to re-send DHCPREQUEST, etc...
So yes, even in the successful case, you could tell how many attempts it took for dhclient to actually negotiate the dhcp leases successfully. Ideally we would want to see a timestamp PER message of dhclient, but this is only available through journalctl (unless we stream the stderr pipe to cloud-init.log, which is going to be not trivial how the current code is setup)

Actually we should log this in both the failed path and the success path. @Moustafa-Moustafa  please ensure we log this in both failed and success path. My understanding is that we already have logging in the failed path, or is that not true?

In fact, if we can come up with a way to bubble up the stderr back to the caller, we can report a diagnostic event that captures the dhcp output to the platform through the HyperV handler, that'd be ideal.

Reply:
I know this PR was landed months ago, and this comment will probably be ignored, but I would like to question the choice of the word "dhclient *error* stream" that even on a successful DHCP lease will printout "error". I got this while running some tests and left us with some question in mind.